"Parse photos from public group" scenario

wkejqb opened the pull request:
Implement "Parse photos from public group" scenario:
 - refactored existing scenarios, moved those to src/;
 - find group by its username;
 - filter messages between timestamps and return objects with timestamp;
 - added trace generator; scenario integration test using trace.

anton-postuf commented on src/MTSerialization/AnonymousMessage.php:
```diff
@@ -29,6 +31,8 @@ public function getNodes(string $name);
     public function getType();
 
+    public function hasNode(string $name): bool;
+
```
Отходим от концепции fail-fast: при изменении TL-модели можем не узнать об этом своевременно

wkejqb replied:
убрал

anton-postuf commented on src/Scenario/GroupPhotosClient.php:
```diff
@@ -0,0 +1,221 @@
+        usleep(100000);
+        $this->infoClient->readPicture($locationRequest, $dcId, function (PictureModel $model) use ($id) {
+            $filename = "$id.".$model->format;
+            file_put_contents($filename, $model->bytes);
```
Подобные вещи необходимо параметризировать

wkejqb replied:
добавил параметр-функцию.

wkejqb commented on src/Client/InfoObtainingClient/InfoClient.php:
```diff
@@ -101,6 +102,14 @@ public function getChatMembers(int $id, callable $onComplete) {
+            $onComplete
+        );
+    }
+
```
сделаю это в отдельном MR к следующим сценариям, чтобы не перегружать этот MR.

wkejqb commented on src/Client/InfoObtainingClient/InfoClient.php:
```diff
@@ -286,7 +295,7 @@ private function createUserStatusModel($userStatus)
      * @noinspection PhpDocRedundantThrowsInspection
      */
-    private function readPicture(TLClientMessage $fileLocation, int $photoDcId, callable $onPictureLoaded)
+    public function readPicture(TLClientMessage $fileLocation, int $photoDcId, callable $onPictureLoaded)
```
поправил

wkejqb commented on examples/parsePhotos.php:
```diff
@@ -0,0 +1,35 @@
+        die();
+    } elseif ($argv[1] !== '--info') {
+        $groupId = (int) $argv[1];
+    }
```
сделано

wkejqb commented on src/Client/MultiClient.php:
```diff
@@ -23,19 +27,22 @@ class MultiClient
      * @throws TGException
      */
-    public function __construct(array $authKeysSerialized)
+    public function __construct(array $authKeysSerialized, ?callable $clientCreator = null)
```
Добавил комментарий в phpdoc.
Как вариант, можно использовать классы-генераторы (это я в следующем MR с тестом покажу), но по мне и так неплохо, это не влияет на регулярное использование класса.

wkejqb commented on examples/parsePhotos.php:
```diff
@@ -0,0 +1,47 @@
+Usage: php parsePhotos.php [groupId|deepLink] [dateFrom] [dateTo] [--info]
+    deepLink ex.: https://t.me/vityapelevin
+    dateFrom/dateTo format: YYYYMMdd
+TXT;
```
https://github.com/getopt-php/getopt-php

wkejqb replied:
https://github.com/getopt-php/getopt-php можно взять

wkejqb commented on examples/parsePhotos.php:
```diff
@@ -0,0 +1,47 @@
+$photosClient = new GroupPhotosClient($groupId, $since, $to);
+if ($deepLink) {
+    $photosClient->setDeepLink($deepLink);
+}
```
Сделал для id тоже сеттер, для единообразия, а из конструктора убрал.

wkejqb commented:
Commit message:
```
Implement "Parse photos from public group" scenario:
 - refactored existing scenarios, moved those to src/;
 - added trace generator; scenario integration test using trace.
```
